pyani_plus/tools.py

```diff
@@ -0,0 +1,197 @@
+#
+# Peter Cock,
+# Strathclyde Institute of Pharmaceutical and Biomedical Sciences
+# The University of Strathclyde
+# 161 Cathedral Street
+# Glasgow
+# G4 0RE
+# Scotland,
+# UK
+#
+# The MIT License
+#
+# (c) The University of Strathclyde 2024-present
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""Assorted functions for capturing command line tool version information.
+
+The pyANI-plus software relies on multiple separate command line tools
+such as ``nucmer`` from mummer3 to computer average nucleotide identity.
+This module contains helper functions to get the version of such tools,
+typically by executing them and parsing the stdout. These can be used for
+logging, as the exact versions used can sometimes be important for full
+reproducibility of results.
+"""
+
+import re
+import shutil
+import subprocess
+from pathlib import Path
+from typing import NamedTuple
+
+
+class ExternalToolData(NamedTuple):
+    """Convenience struct for tool path and version information."""
+
+    exe_path: Path
+    version: str
+
+
+def check_cmd(cmd: str | Path) -> Path:
+    """Return full Path of given command, or raise RuntimeError."""
+    if not cmd:
+        msg = "Function check_cmd requires a command or full path."
+        raise ValueError(msg)
+
+    exe_str = shutil.which(cmd)
+    if not exe_str:
+        msg = (
+            f"{cmd} is not executable"
+            if "/" in str(cmd)
+            else f"{cmd} not found on $PATH"
+        )
+        raise RuntimeError(msg)
+
+    # Turn into an absolute path in case later use a temp working dir:
+    return Path(exe_str).absolute()
+
+
+def _get_path_and_version_output(
+    cmd: str | Path, args: list[str] | None = None
+) -> tuple[Path, str]:
+    """Determine path of command, run it with args, capture combined stdout and stderr."""
+    # Might later need to add check=True as an optional argument,
+    # e.g. NCBI legacy blast doesn't have a version option and uses return code 1.
```

Comment:
I'd like us to drop legacy BLAST with this version of pyani - but as well to keep in mind that not all tools we might use are well-behaved.